Align diffusers CI tests with examples (#1679)
huggingface/optimum-habana: examples/stable-diffusion/README.md

# Stable Diffusion Examples

> This directory contains sample scripts that showcase how to perform diffusion-based generative tasks on Intel® Gaudi® AI Accelerators.

Reviewer: ```suggestion
This directory contains sample scripts demonstrating how to perform diffusion-based generative tasks on Intel® Gaudi® AI Accelerators.
```

> Stable Diffusion was proposed in [Stable Diffusion Announcement](https://stability.ai/blog/stable-diffusion-announcement) by Patrick Esser
> and Robin Rombach and the Stability AI team.

Reviewer: ```suggestion
Stable Diffusion was introduced in [Stable Diffusion Announcement](https://stability.ai/blog/stable-diffusion-announcement) by Patrick Esser, Robin Rombach and the Stability AI team.
```

pip install -r requirements.txt
```

## Text-to-Image Generation

### Stable Diffusion

> Here is how to generate images using Stable Diffusion 1.4 model with one prompt:

Reviewer: ```suggestion
Here's how to generate images using the Stable Diffusion 1.4 model with a single prompt:
```

    --bf16
```

> [!NOTE]
> HPU graphs are recommended when generating images by batches to get the fastest possible generations.
> The first batch of images entails a performance penalty. All subsequent batches will be generated much faster.
> You can enable this mode with `--use_hpu_graphs`.

> To generate images with several prompts, simply include 2 prompts in your input as follows:

Reviewer: ```suggestion
To generate images with multiple prompts, simply include two prompts in your input as shown below:
```

    --bf16
```

> Distributed inference with multiple HPUs is also supported. Here is an example which shows
> how to generate images with two prompts on two HPUs:

Reviewer: ```suggestion
Distributed inference with multiple HPUs is also supported. Below is an example demonstrating how to generate images with two prompts on two HPUs:
```

    --distributed
```

> You can run other older Stable Diffusion models in a similar manner. For example, to perform image generation with
> Stable Diffusion 1.5, use the option: `--model_name_or_path stable-diffusion-v1-5/stable-diffusion-v1-5`. Examples
> showcasing Stable Diffusion 2 are given next.

Reviewer: ```suggestion
You can run other older Stable Diffusion models in a similar manner. For example, to generate images with Stable Diffusion 1.5, use the option: `--model_name_or_path stable-diffusion-v1-5/stable-diffusion-v1-5`. Examples showcasing Stable Diffusion 2 are provided next.
```

    --bf16
```

> [!NOTE]
> HPU graphs are recommended when generating images by batches to get the fastest possible generations.
> The first batch of images entails a performance penalty. All subsequent batches will be generated much faster.
> You can enable this mode with `--use_hpu_graphs`.

> SDXL incorporates a second text encoder (OpenCLIP ViT-bigG/14) alongside the original Stable Diffusion text encoder,
> significantly increasing the number of parameters and enabling more detailed and descriptive prompts. Here is an
> example of how to generate images using multiple prompts for both `prompt` (primary text encoder) and `prompt_2`
> (secondary text encoder), along with their respective negative prompts:

Reviewer: ```suggestion
SDXL integrates a second text encoder (OpenCLIP ViT-bigG/14), alongside the original Stable Diffusion text encoder. This addition significantly increases the number of parameters, enabling more detailed and descriptive prompts. Below is an example of how to generate images using multiple prompts for both `prompt` (primary text encoder) and `prompt_2` (secondary text encoder), along with their respective negative prompts:
```

    --bf16
```

> Distributed inferencing with Intel Gaudi accelerators is also supported for SDXL.  Here is an example of generating
> SDXL images in a distributed manner using two prompts on two HPUs:

Reviewer: ```suggestion
SDXL also supports distributed inferencing with Intel Gaudi accelerators. Below is an example of generating SDXL images in a distributed manner using two prompts on two HPUs:
```

    --distributed
```

> The performance-optimized SDXL pipeline can be enabled using the `--optimize` option. This option leverages a more
> aggressively optimized modified attention mechanism for improved performance. Additionally, it supports running

Reviewer: ```suggestion
The performance-optimized SDXL pipeline can be enabled using the `--optimize` option. This option utilizes a more aggressively optimized attention mechanism for enhanced performance. Additionally, it supports running
```

inference in mixed FP8 precision.

> Here is how to generate SDXL images with optimized pipeline in FP8 precision:

Reviewer: @dsocek this command fails!

```sh
    raise FileNotFoundError(f"Failed to load file {fname}")
FileNotFoundError: Failed to load file ./quantization/measure/fp8_hooks_maxabs.npz
```

Maybe we need to add a note or st like below

```txt
**Note:** Before executing the command below, please measure the model quantization statistics.
```

Reviewer: or maybe referring to `QUANT_CONFIG=./quantization/stable-diffusion-xl/measure_config.json`

Author: Good catch.  I seem to forgot to update quantization `json` files with updated path after I moved them to sdxl subfolder.   Fixed now.

    --optimize
```

#### SDXL-Turbo

> Knowledge distillation technique can be used to train a distilled version of SDXL, where fewer
> inference steps are sufficient to generate high-quality images. SDXL-Turbo is a distilled
> version of Stable Diffusion XL 1.0, trained for real-time synthesis.

Reviewer: ```suggestion
The knowledge distillation technique can be used to train a distilled version of SDXL, allowing for high-quality image generation with fewer inference steps. SDXL-Turbo is a distilled version of Stable Diffusion XL 1.0, optimized for real-time synthesis.
```

## ControlNet

> Here is how to generate images conditioned by the canny edge model:

Reviewer: ```suggestion
```

Reviewer: repeated a couple of lines below

ControlNet was introduced in [Adding Conditional Control to Text-to-Image Diffusion Models](https://huggingface.co/papers/2302.05543)

> by Lvmin Zhang and Maneesh Agrawala. It enables conditioning the Stable Diffusion model with an additional input image, allowing
> for precise control over the composition of the generated images using various features such as edges, pose, depth, and more.

Reviewer: ```suggestion
by Lvmin Zhang and Maneesh Agrawala. enables conditioning the Stable Diffusion model with an additional input image. This allows for precise control over the composition of generated images using various features such as edges, pose, depth, and more.
```

    --bf16
```

The ControlNet example can be run with multiple prompts by supplying more than one prompt in the input.

> In addition, it can also be run in a distributed manner. Here is how to generate images conditioned by
> Canny edge model and with two prompts on two HPUs:

Reviewer: ```suggestion
Additionally, it supports distributed execution. Below is an example of generating images conditioned by the Canny edge model using two prompts on two HPUs:
```

    --distributed
```

> These ControlNet examples will also do preprocessing of the input image and derive Canny edges from it. More generally,
> we can use `--control_preprocessing_type none` and supply preprocessed control image directly which allows running many 
> additional use cases.

Reviewer: ```suggestion
These ControlNet examples will preprocess the input image to derive Canny edges. Alternatively, you can use `--control_preprocessing_type none` to supply a preprocessed control image directly, enabling many additional use cases.
```

    --bf16
```

## Additional Stable Diffusion-based Inference Techniques

> This section provides examples demonstrating additional inference techniques based on Stable Diffusion.
> For more details, please refer to

Reviewer: ```suggestion
This section provides examples of additional inference techniques based on Stable Diffusion. For more details, please refer to
```

[Hugging Face Diffusers documentation](https://huggingface.co/docs/diffusers/main/en/using-diffusers/overview_techniques).

### Unconditional Image Generation

> Here is how to perform unconditional image generation on Intel Gaudi. For more details, please see

Reviewer: ```suggestion
Here is how to perform unconditional image generation on Intel Gaudi. For more details,  please refer to the 
```

[Unconditional Image Generation](https://huggingface.co/docs/diffusers/using-diffusers/unconditional_image_generation)
section in the Hugging Face documentation.

```bash
python unconditional_image_generation.py \
    --model_name_or_path "google/ddpm-ema-celebahq-256" \
    --batch_size 16 \
    --use_habana \
    --use_gaudi_ddim_scheduler \
    --use_hpu_graphs \
    --sdp_on_bf16 \
    --bf16 \
    --save_outputs \
    --output_dir "/tmp/"
```

### Controlling Brightness

> Here is how to run an example of controlling brightness. For more information, please refer to the

Reviewer: ```suggestion
Here is an example of how to control brightness. For more information, please refer to the
```

[Control Brightness](https://huggingface.co/docs/diffusers/main/en/using-diffusers/control_brightness)
section in the Hugging Face documentation.

```bash
PT_HPU_MAX_COMPOUND_OP_SIZE=1 \
python text_to_image_generation.py \
    --model_name_or_path ptx0/pseudo-journey-v2 \
    --prompts "A lion in galaxies, spirals, nebulae, stars, smoke, iridescent, intricate detail, octane render, 8k" \
    --num_images_per_prompt 1 \
    --batch_size 1 \
    --use_habana \
    --use_hpu_graphs \
    --image_save_dir /tmp/stable_diffusion_images_brightness \
    --seed 33 \
    --use_zero_snr \
    --guidance_scale 0.7 \
    --timestep_spacing trailing
```

### Prompt Weighting

> Here is how to run an example of prompt weighting. For more information, please refer to the

Reviewer: ```suggestion
Here is an example of how to run prompt weighting. For more information, please refer to the
```

[Weighted Prompts](https://huggingface.co/docs/diffusers/main/en/using-diffusers/weighted_prompts)
section in the Hugging Face documentation.

```bash
python text_to_image_generation.py \
    --model_name_or_path CompVis/stable-diffusion-v1-4 \
    --prompts "a red cat playing with a ball+++" "a red cat playing with a ball---" \
    --num_images_per_prompt 4 \
    --batch_size 4 \
    --use_habana --use_hpu_graphs \
    --image_save_dir /tmp/stable_diffusion_images_compel \
    --seed 33 \
    --sdp_on_bf16 \
    --bf16 \
    --num_inference_steps 20 \
    --use_compel
```

### Controlling Image Quality

> Here is how to run an example of improving image quality. For more details, please see

Reviewer: ```suggestion
Here is an example of how to improve image quality. For more details, please see
```

[Image Quality](https://huggingface.co/docs/diffusers/main/en/using-diffusers/image_quality)
section in the Hugging Face documentation.

```bash
python text_to_image_generation.py \
    --model_name_or_path CompVis/stable-diffusion-v1-4 \
    --prompts "A squirrel eating a burger" \
    --num_images_per_prompt 4 \
    --batch_size 4 \
    --use_habana \
    --image_save_dir /tmp/stable_diffusion_images_freeu \
    --seed 33 \
    --use_freeu \
    --sdp_on_bf16 \
    --bf16
```

## Image-to-Image Generation

> Images can also be generated from initial input images that guide the diffusion-based image generation process

Reviewer: ```suggestion
Images can also be generated using initial input images to guide the diffusion-based image generation process.
```

### Stable Diffusion-based Image-to-Image

> Here is how to generate images with one prompt and one image using `timbrooks/instruct-pix2pix` model which is based on the Stable Diffusion:

Reviewer: ```suggestion
Here is how to generate images using a single prompt and an input image with the `timbrooks/instruct-pix2pix` model, which is based on Stable Diffusion:
```

    --bf16
```

> [!NOTE]
> HPU graphs are recommended when generating images by batches to get the fastest possible generations.
> The first batch of images entails a performance penalty. All subsequent batches will be generated much faster.
> You can enable this mode with `--use_hpu_graphs`.

This example can also be run with multiple prompts by supplying more than one prompt in the input.

### Stable Diffusion XL Refiner

> Here is how to refine SDXL images with a single image and prompt:

Reviewer: ```suggestion
Here is how to refine SDXL images using a single image and prompt:
```

    --bf16
```

### FLUX.1 Image-to-Image

> Here is how to generate FLUX.1 image with one input image and a single prompt:

Reviewer: ```suggestion
Here is how to generate a FLUX.1 image using a single input image and prompt:
```

    --bf16
```

## Text-to-Video Generation

> This section showcases how to use the `GaudiTextToVideoSDPipeline` to run text-to-video generation tasks on HPUs.
> The pipeline uses a UNet3D structure and implements video generation through the iterative denoising process.

Reviewer: ```suggestion
This section demonstrates how to use the `GaudiTextToVideoSDPipeline` for text-to-video generation tasks on HPUs.
The pipeline employs a UNet3D structure and generates videos through an iterative denoising process.
```

```bash

> python3 text_to_video_generation.py \

Reviewer: ```suggestion
python text_to_video_generation.py \
```
